### PR TITLE
reproducibility PR1 : introduce ensembler 

### PR DESCRIPTION
**Pull request by shahules786**
Changes

- [x] An ensemble to combine multiple outputs for the same query (binary)
- [x] introduce the concept of reproducibility to 3 metrics (self consistency)
- [x] Modify and break answer-correctness into 2 step process to improve stability and decrease errors.

#925

### CODE REVIEW

File: src/ragas/metrics/_context_precision.py [hunk 1]
```diff
@@ -90,6 +90,18 @@ class ContextPrecision(MetricWithLLM):
     evaluation_mode: EvaluationMode = EvaluationMode.qcg  # type: ignore
     context_precision_prompt: Prompt = field(default_factory=lambda: CONTEXT_PRECISION)
     max_retries: int = 1
+    _reproducibility: int = 1
+
+    @property
+    def reproducibility(self):
+        return self._reproducibility
```
Comment (jjmachan) [inline src/ragas/metrics/_context_precision.py:97]: shouldn't this be part of `Metrics`?

File: src/ragas/metrics/base.py [hunk 2]
```diff
@@ -156,3 +157,30 @@ def init(self, run_config: RunConfig):
...
+
+        verdict_agg = []
+        for i in range(len(inputs[0])):
+            item = inputs[0][i]
+            verdicts = [inputs[k][i][attribute] for k in range(len(inputs))]
+            verdict_counts = dict(Counter(verdicts).most_common())
+            item[attribute] = list(verdict_counts.keys())[0]
+            verdict_agg.append(item)
+
+        return verdict_agg
```
Comment (jjmachan) [inline src/ragas/metrics/base.py:183]: how does this work?

File: src/ragas/metrics/_context_precision.py (hunk 1 above)
Comment (jjmachan) [reply src/ragas/metrics/_context_precision.py:97]: The question you have to ask is - does every `Metric` or `MetricWithLLM` or `MetricWithEmbedding` need this?

and you only have to ask the question if you have to do this 2+ times - which is this case you did. so what was the consideration here?

**Review by jjmachan — CHANGES_REQUESTED** [review]

File: src/ragas/metrics/_context_precision.py (hunk 1 above)
Comment (shahules786) [reply src/ragas/metrics/_context_precision.py:97]: I did think about this. For now, only 3/6 of the metrics can use this technique. I tried multiple ways to combine generative o/p too for improved reproducibility but I didn't get satisfactory results to raise a PR.  So, given that only 3/6 metrics have this wouldn't it be okay to keep it here now?

File: src/ragas/metrics/base.py (hunk 2 above)
Comment (shahules786) [reply src/ragas/metrics/base.py:183]: I just ensembles multiple outputs. 
For example if results is [0,1,1] over 3 llm calls for the query, output will be 1 which is the majority vote. I will add a doc for this.

**Review by jjmachan — APPROVED** [review]

**Pull request merged**